### ISSUE TITLE
Uncaught TypeError: Cannot read properties of undefined (reading 'substring')

### ISSUE THREAD

**Issue by NeilFraser** [ajaxorg/ace#4583]
Google's Blockly Games pulls in a compressed version of ACE from:
https://github.com/ajaxorg/ace-builds/trunk/src-min-noconflict/

We are logging client-side errors and periodically observe this one:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'substring')
https://blockly.games/third-party/ace/worker-javascript.js 1:7931
```

This matches the last 'substring' on [these lines](https://github.com/ajaxorg/ace-builds/blob/d979a82c6198802cfb6df5be236379d29f116b2d/src-noconflict/worker-javascript.js#L535):
```
        case "remove":
            var endColumn = delta.end.column;
            var endRow = delta.end.row;
            if (row === endRow) {
                docLines[row] = line.substring(0, startColumn) + line.substring(endColumn);
            } else {
                docLines.splice(
                    row, endRow - row + 1,
                    line.substring(0, startColumn) + docLines[endRow].substring(endColumn)
                );
            }
```

Thus, `docLines[endRow]` is sometimes undefined.  I don't have any information about the state of the user's editors or what they were doing at the time this error was thrown.  All browsers (Chrome, Edge, Safari), all operating systems (OS X, ChromeOS and Windows).

If there's nothing that can be figured out from your side, I'm happy to inject some instrumentation into our compiled package to tease out more info on what's happening.  This error pops up about once every two hours, so it won't take long.

**Comment by nightwing** [comment]
It seems like the change information sent from 
[worker_client.js#L198](https://github.com/ajaxorg/ace/blob/master/lib/ace/worker/worker_client.js#L198) through
[worker_client.js#L205](https://github.com/ajaxorg/ace/blob/master/lib/ace/worker/worker_client.js#L205)
 somehow gets desynchronised and [mirror.js#L26 ](https://github.com/ajaxorg/ace/blob/master/lib/ace/worker/mirror.js#L26) receives change information that is meant to be applied to a larger document. 
Unfortunately i was not able to find the root cause or reproduce the issue, so additional information would be very useful.

If we had log of arguments passed to postMessage at  https://github.com/ajaxorg/ace/blob/master/lib/ace/worker/worker_client.js#L165 and https://github.com/ajaxorg/ace/blob/master/lib/ace/worker/worker_client.js#L174 it would help with debugging.

**Comment by NeilFraser** [comment]
Progress.  I can now recreate the error.  And I've discovered a second error.

* To recreate, go to: https://blockly.games/pond-duck
* Click the 'JavaScript' tab.
* Triple-click the 'cannon(0, 70);' code so that everything is selected (critically including the line feed at the end).
* Copy.
* Paste.
* Press cancel twice.

To errors show up.  First, the original one that started all this:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'substring')
    at exports.applyDelta (worker-javascript.js:544)
    at Document.applyDelta (worker-javascript.js:1104)
    at Array.<anonymous> (worker-javascript.js:1387)
    at Sender.EventEmitter._signal (worker-javascript.js:597)
    at window.onmessage (worker-javascript.js:201)
```

This is the result of calling `editor.setValue(code, -1);` when cancel is pressed.  Now that I can see this error myself, I'll keep working on figuring out what's going on and why there are two cancel events.

Second, an error in ACE's error trap:
```
Uncaught TypeError: Cannot read properties of null (reading 'data')
    at window.onerror (worker-javascript.js:25)
```

For the second error, logging confirms that `err` is null.  I assume that this one is easy enough to fix on your end.

**Comment by NeilFraser** [comment]
I'm now understanding what's happening.

We have an event handler that fires on every editor change:
```
  session['on']('change', editorChanged);
```
This handler sometimes chooses to revert the most recent edit:
```
function editorChanged() {
  editor.setValue(previousValue, -1);
}
```
It turns out that adding a new line (using the keyboard) and instantly reverting it (using setValue) leaves worker-javascript.js in an ugly state.

I've apparently resolved the issue on my end by adding a 0ms delay:
```
function editorChanged() {
  setTimeout(function() {
    editor.setValue(previousValue, -1);
  }, 0);
}
```

My interpretation is when the user commands a delta the event call is made first, then worker-javascript.js is being told to update based on the delta, and if the event call changes the text then the worker gets a delta which doesn't match the current editor state.

The lesson is that the event handler should not call setValue synchronously.  Not sure if there's anything you should do in the ACE codebase to guard against this, or if it's a documentation issue.

And there's also the second issue that the error trap in worker-javascript.js is broken.

I'll go update, recompile, and push a new version of Blockly Games to resolve this issue from my end with a setTimeout.  Let me know if you need any more info from me.

**Comment by NeilFraser** [comment]
Slight complication with the second issue.  The error trap appears to work fine when tested online (https:// protocol), but fails when tested locally (file:// protocol).

**Comment by andrewnester** [comment]
I've added a PR to fix the issue on Ace side with some explanation on why it happens
https://github.com/ajaxorg/ace/pull/4831

@NeilFraser thanks a lot for your help and details for this issue, it was very useful

**Comment by andrewnester** [comment]
Fixed and released in 1.7.1
